enterprise/server/remote_execution/nbd/nbdclient/nbdclient.go

```diff
@@ -27,8 +30,184 @@ const (
+// goroutines running in the same binary. For example, if all goroutines enter a
+// syscall which depends on an NBD read (such as a read() or execve()), then it
+// will lead to deadlock, because the nbd client goroutine would not be able to
+// run.
+type Process struct {
+	name   string
+	input  *os.File
+	output *bufio.Scanner
+	err    error
+}
+
+// Start starts an nbdclient as a background process. The returned object can be
+// used to communicate with the process.
+func Start(name, mountPath string) (*Process, error) {
+	inR, inW, err := os.Pipe()
+	if err != nil {
+		return nil, status.InternalErrorf("failed to create pipe: %s", err)
+	}
+	outR, outW, err := os.Pipe()
+	if err != nil {
+		return nil, status.InternalErrorf("failed to create pipe: %s", err)
+	}
+	cmd := exec.Command(os.Args[0], append(os.Args[1:], "--nbdclient")...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.ExtraFiles = []*os.File{inR, outW}
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	// Close input reader and output writer handles; these are only used by the
+	// child.
+	inR.Close()
+	outW.Close()
+
+	scanner := bufio.NewScanner(outR)
+	h := &Process{name: name, input: inW, output: scanner}
+	go func() {
+		err := cmd.Wait()
+		h.err = status.InternalErrorf("child exited: %s", err)
+		log.Error(h.err.Error())
+		inW.Close()
+		outR.Close()
+	}()
+
+	if err := h.Init(name, mountPath); err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+// Init initializes the ClientDevice in the client process.
+func (h *Process) Init(name, mountPath string) error {
+	return h.send(initCmd, name, mountPath)
+}
+
+// Mount mounts the ClientDevice in the client process.
+func (h *Process) Mount() error {
+	return h.send(mountCmd)
+}
+
+// Unmount unmounts the ClientDevice in the client process.
+func (h *Process) Unmount() error {
+	return h.send(unmountCmd)
+}
+
+func (h *Process) send(cmd string, args ...string) error {
+	msg := strings.Join(append([]string{cmd}, args...), " ")
+	log.Infof("nbdclient %s: sending control message %q", h.name, msg)
+	if _, err := h.input.WriteString(msg + "\n"); err != nil {
+		return err
+	}
+	if !h.output.Scan() {
+		return status.WrapError(h.err, "recv")
+	}
+	if reply := h.output.Text(); reply != "OK" {
+		return status.InternalErrorf("unexpected reply: %s", reply)
+	}
+	log.Infof("nbdclient %s: OK", h.name)
+	return nil
+}
+
+// Run is the background process implementation for the process spawned by
+// StartProcess. The init binary checks the "--nbdclient" flag and passes
+// control to this function if present.
+func Run(ctx context.Context) error {
+	ctx = log.EnrichContext(ctx, "pid", fmt.Sprint(os.Getpid()))
+	log.CtxInfof(ctx, "nbdclient child process started")
+	// fds 3 and 4 are the files passed to us via ExtraFiles.
+	in := os.NewFile(3, "in")
+	out := os.NewFile(4, "out")
+	scanner := bufio.NewScanner(in)
+
+	var device *ClientDevice
+	var mountPath string
+
+	for scanner.Scan() {
```

Comment:
I might be misunderstanding how this works, but could there be a race condition if we haven't sent commands to this background process fast enough? i.e. If the input FD hasn't received any commands yet, then scanner.Scan would return immediately

Reply:
Linux pipes work similarly to buffered channels in Go - the `read()` syscall that underlies `scanner.Scan()` will block until there is data available in the pipe buffer, and `write()` on the other end of the pipe will either add to the pipe buffer, or if the buffer is full, it waits until there's available capacity.